pkg/queue/queue_test.go

```diff
@@ -333,6 +336,208 @@ func TestMaxQueueSize(t *testing.T) {
+			tenantsCount:          100,
+			tasksPerTenant:        10,
+			consumersCount:        10,
+			dequeueBatchSize:      6,
+			maxConsumersPerTenant: 4,
+		},
+		"tenants-200_tasks-10_consumers-100_batch-999999_maxConsumersPerTenant-4": {
+			tenantsCount:          200,
+			tasksPerTenant:        10,
+			consumersCount:        100,
+			dequeueBatchSize:      999_999,
+			maxConsumersPerTenant: 4,
+		},
+		"tenants-10_tasks-100_consumers-20_batch-5_maxConsumersPerTenant-16": {
+			tenantsCount:          10,
+			tasksPerTenant:        100,
+			consumersCount:        20,
+			dequeueBatchSize:      5,
+			maxConsumersPerTenant: 16,
+		},
+		"tenants-1_tasks-100_consumers-16_batch-5_maxConsumersPerTenant-2": {
+			tenantsCount:          1,
+			tasksPerTenant:        100,
+			consumersCount:        16,
+			dequeueBatchSize:      5,
+			maxConsumersPerTenant: 2,
+		},
+	}
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			totalTasksCount := tt.tenantsCount * tt.tasksPerTenant
+			limits := &mockLimits{maxConsumer: tt.maxConsumersPerTenant}
+			require.LessOrEqual(t, tt.tasksPerTenant, tenantsQueueMaxSize, "test must be able to enqueue all the tasks without error")
+			queue := NewRequestQueue(tenantsQueueMaxSize, 0, limits, NewMetrics(nil, constants.Loki, "query_scheduler"))
+			for i := 0; i < tt.consumersCount; i++ {
+				queue.RegisterConsumerConnection(createConsumerName(i))
+			}
+
+			ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+			defer cancel()
+			wg, _ := errgroup.WithContext(ctx)
+
+			mtx := &sync.Mutex{}
+			receivedTasksPerTenant := make(map[string]int, tt.tenantsCount)
+			receivedTasksCount := atomic.NewInt32(0)
+			for i := 0; i < tt.consumersCount; i++ {
+				consumer := createConsumerName(i)
+				wg.Go(func() error {
+					idx := StartIndexWithLocalQueue
+					for {
+						tasks, newIdx, err := queue.DequeueMany(ctx, idx, consumer, tt.dequeueBatchSize)
+						receivedTasksCount.Add(int32(len(tasks)))
+
+						if err != nil && !errors.Is(err, context.Canceled) {
+							return fmt.Errorf("error while dequeueing many task by %s: %w", consumer, err)
+						}
+						if err == nil && len(tasks) == 0 {
+							return fmt.Errorf("%s must receive at least one item if there is no error, but got 0", consumer)
+						}
+						if len(tasks) > 1 && !isAllItemsSame(tasks) {
+							return fmt.Errorf("expected all items to be from the same tenant, but they were not: %v", tasks)
+						}
+						time.Sleep(tt.consumerDelay)
+						mtx.Lock()
+						if err == nil {
+							//tenant name is sent as task
+							tenant := tasks[0]
+							receivedTasksPerTenant[fmt.Sprintf("%v", tenant)] += len(tasks)
+						}
+						// put the slice back to the pool
+						queue.ReleaseRequests(tasks)
+						if receivedTasksCount.Load() == int32(totalTasksCount) {
+							//cancel the context to release the rest of the consumers once we received all the tasks from all the queues
+							cancel()
```

Comment:
I don't really understand why we need to call `cancel()` here.
`cancel()` is deferred on line 641, and I don't see the `ctx.Err()` checked anywhere.

Reply:
as long as we do not know how many iterations we have, each worker will run this `for` to dequeueMany until the context is cancelled... so, once the counter `receivedTasksCount` reaches the expected number, we need to cancel the context to stop the rest of the goroutines... 
`defere cancel()` is here just to avoid leaking goroutines in case the tests fail. 

and `ctx.Err()` is checked inside the queue.